File: Core/Bridge/io/Keyboard.cs
using Godot;
using System;
using System.Collections.Generic;

namespace monoe.exe.Core.Bridge.io;

public static class KeyResolver
{
  private static readonly Dictionary<string, Key> _map;

  static KeyResolver()
  {
    _map = [];

    foreach (Key k in Enum.GetValues<Key>())
    {
      string name = Normalize(k.ToString());
      _map[name] = k;
    }

    Alias("space", Key.Space);
    Alias("esc", Key.Escape);
    Alias("escape", Key.Escape);
    Alias("enter", Key.Enter);
    Alias("return", Key.Enter);
    Alias("left", Key.Left);
    Alias("right", Key.Right);
    Alias("up", Key.Up);
    Alias("down", Key.Down);
  }

  private static void Alias(string name, Key key)
      => _map[Normalize(name)] = key;

  private static string Normalize(string s)
      => s.Replace("_", "")
          .Replace("-", "")
          .Replace(" ", "")
          .ToLowerInvariant();

  public static bool TryResolve(string input, out Key key)
      => _map.TryGetValue(Normalize(input), out key);
}

/// <summary>
/// Primitive-only keyboard input bridge.
/// Exposes physical keys and Godot actions using strings and booleans.
/// </summary>
public static class Keyboard
{
  /// <summary>
  /// Returns true if a physical key is currently pressed.
  /// Example keys: "a", "space", "enter", "esc", "left", "right"
  /// </summary>
  public static bool KeyPressed(string key)
  {
    return KeyResolver.TryResolve(key, out var k)
      && Input.IsPhysicalKeyPressed(k);
  }

  /// <summary>
  /// Returns true if a Godot input action is currently pressed.
  /// </summary>
  public static bool ActionPressed(string action)
    => Input.IsActionPressed(action);

  /// <summary>
  /// Returns true if a Godot input action was just pressed this frame.
  /// </summary>
  public static bool ActionJustPressed(string action)
    => Input.IsActionJustPressed(action);

  /// <summary>
  /// Returns true if a Godot input action was just released this frame.
  /// </summary>
  public static bool ActionJustReleased(string action)
    => Input.IsActionJustReleased(action);

  /// <summary>
  /// Returns the strength of a Godot input action (0.0 → 1.0).
  /// Useful for analog input.
  /// </summary>
  public static double ActionStrength(string action)
    => Input.GetActionStrength(action);

  /// <summary> True if Shift is pressed </summary>
  public static bool Shift() => Input.IsKeyPressed(Key.Shift);

  /// <summary> True if Ctrl is pressed </summary>
  public static bool Ctrl() => Input.IsKeyPressed(Key.Ctrl);

  /// <summary> True if Alt is pressed </summary>
  public static bool Alt() => Input.IsKeyPressed(Key.Alt);

  /// <summary> True if Meta / Super is pressed </summary>
  public static bool Meta() => Input.IsKeyPressed(Key.Meta);
}

File: monoelib/io/keyboard.lua
local engine = require("monoelib.engine")

monoe = monoe or {}
monoe.io = monoe.io or {}
---@class monoe.io.keyboard
---Provides some helpful functions that allows you to use keyboard in the game.
monoe.io.keyboard = {}

local KEYBOARD = "monoe.exe.Core.Bridge.io.Keyboard"

---Returns true if a physical key is currently pressed.
---@param key string Key name (e.g. "a", "space", "enter", "esc")
---@return boolean
function monoe.io.keyboard.key_down(key)
  return engine.staticcall(KEYBOARD, "KeyPressed", key)
end

---Returns true if Shift is pressed.
---@return boolean
function monoe.io.keyboard.shift()
  return engine.staticcall(KEYBOARD, "Shift")
end

---Returns true if Ctrl is pressed.
---@return boolean
function monoe.io.keyboard.ctrl()
  return engine.staticcall(KEYBOARD, "Ctrl")
end

---Returns true if Alt is pressed.
---@return boolean
function monoe.io.keyboard.alt()
  return engine.staticcall(KEYBOARD, "Alt")
end

---Returns true if an action is currently pressed.
---@param action string
---@return boolean
function monoe.io.keyboard.action_down(action)
  return engine.staticcall(KEYBOARD, "ActionPressed", action)
end

---Returns true if an action was just pressed this frame.
---@param action string
---@return boolean
function monoe.io.keyboard.action_just_down(action)
  return engine.staticcall(KEYBOARD, "ActionJustPressed", action)
end

---Returns true if an action was just released this frame.
---@param action string
---@return boolean
function monoe.io.keyboard.action_just_up(action)
  return engine.staticcall(KEYBOARD, "ActionJustReleased", action)
end

---Returns the strength of an action (0.0 → 1.0).
---@param action string
---@return number
function monoe.io.keyboard.action_strength(action)
  return engine.staticcall(KEYBOARD, "ActionStrength", action)
end

_G.monoe = monoe

return monoe.io.keyboard